### ISSUE TITLE
server: unresolved parallel decode-overlap cancellation and slot-reuse validation

### ISSUE THREAD

**Issue by GenerelSchwerz** [GenerelSchwerz/llama.cpp#82]
## Problem

Parallel HTTP cancellation tests observed different generated tokens with `--decode-overlap` enabled, despite fixed prompts and seeds. The validation completed before merging #79 did not fully distinguish scheduling-related numerical variation from incorrect KV/recurrent state handling during cancellation and admission.

This tracks an unresolved validation gap, not confirmed state corruption. Overlap is opt-in. These observations do not establish a regression with overlap disabled.

## Evidence and scope

PR #79 merged as `078a36bcdedb96e5ec72b774bdb46228b3aa8e55`. The original GPT-OSS/Muse observations below came from its pre-merge sampling-extension candidate, based on `550a1754dc508359b42e3bdbe14f54704469ce20`, not a fresh reproduction at the exact merge commit. Candidate patch SHA256: `305dc7555fafd5c1897da08d7fb2fd2950b909eeaed1a794489146b8753262f4`.

| Tested path | Continuing peer | Reused slot |
| --- | --- | --- |
| GPT-OSS-20B MXFP4, ordinary device-input overlap | Different tokens | Different tokens |
| Muse-Glimmer-30B Q3_K_M, ordinary device-input overlap | Different tokens | Exact match |
| MTP draft overlap, separate later implementation | Exact match | Exact match |

The GPT-OSS and Muse controls explicitly enabled ordinary GPU backend sampling in both modes, so those comparisons did not conflate CPU sampling with GPU sampling. The requests completed and the slots were released; the discrepancy was output equality, not a recorded crash or hang.

For GPT-OSS, instrumentation reconstructed identical committed random draws for the 192-token peer, including checkpoint restores. Ordinary-decode output also changed when its admission schedule changed. This supports scheduling-related numerical variation and rules out a skipped/duplicated peer RNG draw in that instrumented run. It does not rule out all state-handling errors. Controlled backend lifecycle tests passed exact equality, including pending cancellation and long prefixes.

A later, unpushed Flash Next host-input candidate also changed the continuing peer output under concurrent cancellation, for greedy and stochastic sampling. Two repeated disabled controls matched each other. Its reused-slot output matched. That candidate is separate from merged #79 and is now restricted locally to one slot; its parallel root cause remains unresolved. With that guard, parallel fallback matched the properly isolated inherited build exactly. It should not be reported as a confirmed merged Flash Next regression.

MTP cancellation passed exact peer/reuse comparisons for both samplers in the tested runs, but those HTTP schedules were not deterministic either.

## Reproduction workload

Original hardware: RTX 5070 Ti, 16 GiB VRAM, Linux, CUDA Release build. No MTP for the GPT-OSS/Muse comparison.

Start each mode in a fresh process using the same model and options, adding `--decode-overlap` only for the enabled run:

```sh
llama-server -m /path/to/model.gguf \
  --host 127.0.0.1 --port 8081 \
  -c 4096 -b 512 -ub 512 -np 2 -ngl all \
  -ot token_embd.weight=CUDA0 \
  --backend-sampling --no-warmup --slots -lv 5 --log-colors off
```

Use `/completion` with `temperature: 0.8`, `top_k: 40`, `top_p: 0.9`, `min_p: 0.05`, `seed: 12345`, `ignore_eos: true`, `cache_prompt: false`, and `return_tokens: true`.

Construct the base prompt as:

```python
prompt = ("The river flows past the old bridge. " * 180) + "Enumerate 200 detailed steps for validating a small software service. Start with step 1 and continue with one concrete sentence per step."
```

1. Start a streaming request in slot 1 for 192 tokens, using the base prompt plus ` Focus on client-visible behavior.`
2. After receiving its first token, start a streaming request in slot 0 for 512 tokens using the base prompt.
3. Close the slot-0 connection after receiving 12 tokens.
4. Poll `/slots` until slot 0 is idle. Verify slot 1 is still active.
5. Immediately reuse slot 0 with the base prompt for 96 tokens.
6. Compare complete peer and reused-request token arrays between modes. Confirm enabled logs show queued and discarded work for the cancelled slot.

Client timing changes the batch schedule, so this workload demonstrates the symptom but is not a deterministic reproducer of one root cause.

## Required follow-up

- Reproduce against the exact merged implementation with matching backend sampling in both modes.
- Control admission/cancellation at server decode boundaries and compare identical batch schedules.
- Compare KV/recurrent positions, sampler checkpoints, committed random draws, and logits before the first differing token.
- Exercise greedy and stochastic sampling, cancellation with a live peer, immediate reuse, and fallback after new prompt admission.
- If a state error is found, fix it and add coverage in the existing lifecycle tests. If this is numerical variation caused by changed batching, demonstrate it with a matched ordinary-decode control and document the reproducibility boundary.

Until this distinction is established, parallel cancellation should not be described as fully validated merely because requests complete and the controlled backend tests pass.

Local evidence is retained in `decode-overlap-sampling-20260909/REPORT.md`, its `gptoss-final-cancellation`, `muse-final-cancellation`, and `gptoss-rng-trace-cancellation` directories; `decode-overlap-drafting-mtp-20260909/RESULTS.md`; and `flash-next-overlap-qwen4exp-20260909/VALIDATION_NOTES.md`. These are local artifact names, not public attachments. Saved-build comparisons used as independent evidence must isolate their shared libraries as well as executables.

Investigation and report prepared with Codex assistance.
